File: python/basics/trie/208_implement_trie.py
'''
#https://leetcode.com/problems/implement-trie-prefix-tree/description/


prefix tree
each character in a word is a dictionary.
"." to mark the end of the word

'''

class Trie:
    def __init__(self):
        self.trie = {}

    def insert(self, word:str) -> None:
        d = self.trie
        for c in word:
            if c not in d:
                d[c] = {}
            d = d[c]
        d["."] = "."

    def search(self, word:str) -> bool:
        d = self.trie
        for c in word:
            if c not in d:
                return False
            d = d[c]
        return "." in d

    def startsWith(self, prefix:str) -> bool:
        d = self.trie
        for c in prefix:
            if c not in d:
                return False
            d = d[c]
        return True



input = ["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
trie = Trie()
trie.insert("apple")
ret = trie.search("apple")   # return True
print("apple: " + str(ret))
ret = trie.search("app")     # return False
print("app: " + str(ret))
ret = trie.startsWith("app") # return True
print("startsWith app: " + str(ret))
trie.insert("app")
ret = trie.search("app")     # return True
print("app: " + str(ret))
#[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]

#[null, null, true, false, true, null, true]

File: problem_solving/trie/cpp/208_implement_trie.cpp
/*
*/

#include<string>
#include<unordered_map>
#include<iostream>

class TrieNode {
    public:
    std::unordered_map<char, TrieNode*> children;
    bool is_end_of_word;

    TrieNode() : is_end_of_word(false) {
    }

};


class Trie {
    private:
    TrieNode* root;

    public:
    Trie() : root(new TrieNode()) {
    }

    void insert(const std::string& word) {
        TrieNode* current = root;
        for(char c : word) {
            if(current->children.find(c) == current->children.end()) {
                current->children[c] = new TrieNode();
            }
            current = current->children[c];
        }
        current->is_end_of_word = true;
    }

    bool search(const std::string& word) {
        TrieNode* current = root;
        for(char c : word) {
            if(current->children.find(c) == current->children.end()) {
                return false;
            }
            current = current->children[c];
        }
        return current != nullptr && current->is_end_of_word;
    }

    bool startsWith(const std::string prefix) {
        TrieNode* current = root;
        for(char c : prefix) {
            if(current->children.find(c) == current->children.end()) {
                return false;
            }
            current = current->children[c];
        }
        return current != nullptr;
    }

};


int main() {
    Trie* trie = new Trie();
    
    trie->insert("apple");
    std::cout << trie->search("apple") << std::endl;   // return True
    std::cout << trie->search("app") << std::endl;     // return False
    std::cout << trie->startsWith("app") << std::endl; // return True
    trie->insert("app");
    std::cout << trie->search("app") << std::endl;     // return True
    return 0;
}